doc/LAYERS.org

```diff
@@ -0,0 +1,540 @@
+
+#+begin_src emacs-lisp
+  (use-package helm)
+#+end_src
+
+This simply loads Helm. It is essentially equivalent to =(require 'helm)=.
+
+#+begin_src emacs-lisp
+  (use-package helm
+    :defer t)
+#+end_src
+
+This defers the loading of Helm using the auto-load facility and the auto-load
+commands provided by the Helm source code.
+
+#+begin_src emacs-lisp
+  (use-package helm
+    :defer t
+    :init
+    ;; Code to execute before Helm is loaded
+    :config
+    ;; Code to execute after Helm is loaded
+    )
+#+end_src
+
+This form includes code to execute before and after Helm is loaded. The =:init=
+section can be executed immediately, but since Helm is deferred, the =:config=
+section is not executed until after loading, if ever.
+
+#+begin_src emacs-lisp
+  (use-package helm
+    :commands (helm-find-files helm-M-x))
+#+end_src
+
+This creates auto-load references for additional commands, if you find that the
+package author has been slacking.
+
+#+begin_src emacs-lisp
+  (use-package ruby-mode
+    :mode "\\.rb\\'")
+#+end_src
+
+For packages that provide major modes, you can associate file extensions to that
+mode by using the =:mode= keyword. This adds an entry to =auto-mode-alist= and
+an auto-load for =ruby-mode=. Typically this is not required, as =ruby-mode=
+should already be auto-loadable, and the package should associate Ruby files
+with itself already.
+
+Use-package supports heaps of useful keywords. Look at the [[https://github.com/jwiegley/use-package][documentation]] for
+more.
+
+* Anatomy of a layer
+A layer is simply a folder somewhere in Spacemacs' layer search path that
+contains /at least one/ of these files, ordered vaguely by importance.
+
+- =packages.el= :: for specifying which packages the layer uses
+- =extensions.el= :: for specifying extensions
+- =config.el= :: for miscellaneous configuration
+- =funcs.el= :: for defining general utility functions
+- =keybindings.el= :: for defining general keybindings
+
+Additionally, for each extension, there should be a folder
+=<layer>/extensions/<extension>/= containing the source code for that extension.
+Before initializing an extension, Spacemacs will add this folder to the load
+path for you, so that you can treat it as any other package. The only reason
+Spacemacs treats extensions differently is so that it can do this.
+
+** packages.el
+This file should define a variable called =<layer>-packages=, which should be a
+list of all the packages that this layer needs. Additionally, a layer may
+/exclude/ packages through the variable =<layer>-excluded-packages=. This will
+block packages (and, despite the name, extensions) from being installed, even if
+they are included by another layer.
+
+For each included package, you may define one or more of the following
+functions, which are called in order by Spacemacs to initialize the package.
+
+1. =<layer>/pre-init-<package>=
+2. =<layer>/init-<package>=
+3. =<layer>/post-init-<package>=
```

Comment:
To avoid confusion between spacemacs `init` functions and `init` from `use-package` and `use-package` hooks (`:pre-init` and `:post-init`). All `init` function will be renamed to `configure`. Will become mandatory in 0.105.0 as well.